src/Bicep.Core/TypeSystem/DeployTimeConstantVisitor.cs

```diff
@@ -140,13 +128,50 @@ public override void VisitForSyntax(ForSyntax syntax)
             this.Visit(syntax.CloseSquare);
         }
 
-        public override void VisitObjectSyntax(ObjectSyntax syntax)
+        public override void VisitFunctionCallSyntax(FunctionCallSyntax syntax)
         {
-            if (this.bodyType == null)
+            var currentDeployTimeConstantScopeSyntax = this.deployTimeConstantScopeSyntax;
+
+            if (this.model.Binder.GetSymbolInfo(syntax) is FunctionSymbol functionSymbol &&
+                functionSymbol.FunctionFlags.HasFlag(FunctionFlags.RequiresInlining))
             {
-                return;
+                this.deployTimeConstantScopeSyntax = syntax;
             }
 
+            base.VisitFunctionCallSyntax(syntax);
+
+            if (this.errorSyntax is not null)
+            {
+                this.AppendError();
+            }
+
+            
+            this.deployTimeConstantScopeSyntax = currentDeployTimeConstantScopeSyntax;
+        }
+
+        public override void VisitInstanceFunctionCallSyntax(InstanceFunctionCallSyntax syntax)
+        {
+            var currentDeployTimeConstantScopeSyntax = this.deployTimeConstantScopeSyntax;
+
+            if (this.model.Binder.GetSymbolInfo(syntax) is FunctionSymbol functionSymbol &&
+                functionSymbol.FunctionFlags.HasFlag(FunctionFlags.RequiresInlining))
+            {
+                this.deployTimeConstantScopeSyntax = syntax;
+            }
+
+            base.VisitInstanceFunctionCallSyntax(syntax);
+
+            if (this.errorSyntax is not null)
+            {
+                this.AppendError();
+            }
+
+            
+            this.deployTimeConstantScopeSyntax = currentDeployTimeConstantScopeSyntax;
+        }
```

Comment:
The visitor only flags runtime property references, which is not enough. We should also block the runtime function calls at any location that requires a deploy-time constant. I've created https://github.com/Azure/bicep/issues/2622 to track this.

However, as the visitor gets a bit head-scratching, I feel like we should do a refactoring first and separate the logics instead of keeping adding stuffs, which will make it even less readable.

I think we need 3 visitors. The first one will collect all syntaxes that requires deploy-time constants. The second one visits all the collected syntaxes and flags runtime properties and function calls. The third one also visits the collected syntaxes, but it only checks indirect runtime references in variable declarations.

Thoughts?

Reply:
I think splitting into three much simpler visitors absolutely makes sense to me.

Reply:
Sounds good! I'll start the refactoring once I'm done with other bug fixes.